common/persistence/cassandra/errors.go

```diff
@@ -0,0 +1,231 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cassandra
+
+import (
+	"fmt"
+	"math"
+	"reflect"
+	"sort"
+
+	p "go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/persistence/nosql/nosqlplugin/cassandra/gocql"
+)
+
+var (
+	errorPriority = map[reflect.Type]int{
+		reflect.TypeOf(&p.ShardOwnershipLostError{}):             0,
+		reflect.TypeOf(&p.CurrentWorkflowConditionFailedError{}): 1,
+		reflect.TypeOf(&p.ConditionFailedError{}):                2,
+	}
+
+	errorDefaultPriority = math.MaxInt64
+)
+
+func convertErrors(
+	iter gocql.Iter,
+	requestShardID int32,
+	requestRangeID int64,
+	requestCurrentRunID string,
+	requestRunID string,
+	requestDBVersion int64,
+	requestNextEventID int64, // TODO deprecate this variable once DB version comparison is the default
+) error {
+	errs := extractErrors(
+		iter,
+		requestShardID,
+		requestRangeID,
+		requestCurrentRunID,
+		requestRunID,
+		requestDBVersion,
+		requestNextEventID,
+	)
+	errs = sortErrors(errs)
+
+	if len(errs) == 0 {
+		return nil
+	}
+	return errs[0]
+}
+
+func extractErrors(
+	iter gocql.Iter,
+	requestShardID int32,
+	requestRangeID int64,
+	requestCurrentRunID string,
+	requestRunID string,
+	requestDBVersion int64,
+	requestNextEventID int64, // TODO deprecate this variable once DB version comparison is the default
+) []error {
+
+	var errors []error
+
+	record := make(map[string]interface{})
+	for iter.MapScan(record) {
+		if err := extractShardOwnershipLostError(
+			record,
+			requestShardID,
+			requestRangeID,
+		); err != nil {
+			errors = append(errors, err)
```

Comment:
should you do continue here? or is it possible to have multiple errors from one record?

Reply:
i can do a continue here, since current impl will only be extract one error per iteration, but future impl may not.
plus the error conversion logic will not be frequently invoked